### ISSUE TITLE
Failed to update status for SeldonDeployment "router": Operation cannot be fulfilled on seldondeployments.machinelearning.seldon.io "router": the object has been modified; please apply your changes to the latest version and try again

### ISSUE THREAD

**Issue by pjpan** [SeldonIO/seldon-core#4341]
when I deploy a graph model,  router didn't work exactly. it reports Failed to update status for SeldonDeployment "router" .
----------------------------------------------------------------------------------------
[pengju.pan@t-shhq-aik8s-01 tf-custom-test]$ kc describe seldondeployment router
Name:         router
Namespace:    default
Labels:       <none>
Annotations:  <none>
API Version:  machinelearning.seldon.io/v1
Kind:         SeldonDeployment
Metadata:
  Creation Timestamp:  2022-09-22T11:47:45Z
  Generation:          1
  Managed Fields:
    API Version:  machinelearning.seldon.io/v1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:kubectl.kubernetes.io/last-applied-configuration:
      f:spec:
        .:
        f:annotations:
          .:
          f:seldon.io/rest-timeout:
        f:name:
        f:predictors:
        f:protocol:
    Manager:      kubectl-client-side-apply
    Operation:    Update
    Time:         2022-09-22T11:47:45Z
    API Version:  machinelearning.seldon.io/v1
    Fields Type:  FieldsV1
    fieldsV1:
      f:status:
        .:
        f:address:
          .:
          f:url:
        f:conditions:
        f:deploymentStatus:
          .:
          f:router-test-0-router-crack-detect-gps-detect-image-detect:
            .:
            f:availableReplicas:
            f:replicas:
        f:replicas:
        f:state:
    Manager:         manager
    Operation:       Update
    Time:            2022-09-22T11:48:25Z
  Resource Version:  69739203
  UID:               66969efc-60c1-4e65-ab42-873c238afd6b
Spec:
  Annotations:
    seldon.io/rest-timeout:  25000
  Name:                      routertestr
  Predictors:
    Component Specs:
      Spec:
        Containers:
          Env:
            Name:   NUMBER_OF_ROUTERS
            Value:  3
          Image:    hrb.nioint.com/power/base/seldon/seldonio/custom-router-server:v1.0.0
          Name:     router
    Graph:
      Children:
        Children:
        Env Secret Ref Name:  seldon-rclone-secret
        Implementation:       TENSORFLOW_SERVER_CUSTOM
        Model Uri:            s3://com-nio-pow-data-kuafu-test/seldon_models/battery-surface-crack-cv-cdn
        Name:                 crack-detect
        Parameters:
          Name:   model_type
          Type:   STRING
          Value:  pytorch
        Children:
        Env Secret Ref Name:  seldon-rclone-secret
        Implementation:       TENSORFLOW_SERVER_CUSTOM
        Model Uri:            s3://com-nio-pow-data-kuafu-test/seldon_models/battery-surface-gps-cv-cdn
        Name:                 gps-detect
        Parameters:
          Name:   model_type
          Type:   STRING
          Value:  pytorch
        Children:
        Env Secret Ref Name:  seldon-rclone-secret
        Implementation:       CUSTOM_ALGO_SERVER
        Model Uri:            s3://com-nio-pow-data-kuafu-test/ai-models/battery-surface-image-cv-cdn/v0.0.1
        Name:                 image-detect
      Env Secret Ref Name:    seldon-rclone-secret
      Implementation:         CUSTOM_NODE_SERVER
      Model Uri:              s3://com-nio-pow-data-kuafu-test/seldon_models/inference_graph/router
      Name:                   router
      Type:                   ROUTER
    Name:                     test
    Replicas:                 1
  Protocol:                   seldon
Status:
  Address:
    URL:  http://router-test.default.svc.cluster.local:8000/api/v1.0/predictions
  Conditions:
    Last Transition Time:  2022-09-22T11:48:25Z
    Message:               Deployment has minimum availability.
    Reason:                MinimumReplicasAvailable
    Status:                True
    Type:                  DeploymentsReady
    Last Transition Time:  2022-09-22T11:47:45Z
    Reason:                No HPAs defined
    Status:                True
    Type:                  HpasReady
    Last Transition Time:  2022-09-22T11:47:45Z
    Reason:                No KEDA resources defined
    Status:                True
    Type:                  KedaReady
    Last Transition Time:  2022-09-22T11:47:45Z
    Reason:                No PDBs defined
    Status:                True
    Type:                  PdbsReady
    Last Transition Time:  2022-09-22T11:48:25Z
    Status:                True
    Type:                  Ready
    Last Transition Time:  2022-09-22T11:48:25Z
    Reason:                All services created
    Status:                True
    Type:                  ServicesReady
    Last Transition Time:  2022-09-22T11:48:25Z
    Reason:                All VirtualServices created
    Status:                True
    Type:                  istioVirtualServicesReady
  Deployment Status:
    router-test-0-router-crack-detect-gps-detect-image-detect:
      Available Replicas:  1
      Replicas:            1
  Replicas:                1
  State:                   Available
Events:
  Type     Reason                 Age                From                       Message
  ----     ------                 ----               ----                       -------
  Normal   CreateService          69s                seldon-controller-manager  Created Service "router-test-router"
  Normal   CreateService          69s                seldon-controller-manager  Created Service "router-test-crack-detect"
  Normal   CreateService          69s                seldon-controller-manager  Created Service "router-test-gps-detect"
  Normal   CreateService          69s                seldon-controller-manager  Created Service "router-test-image-detect"
  Normal   CreateDeployment       69s                seldon-controller-manager  Created Deployment "router-test-0-router-crack-detect-gps-detect-image-detect"
  Warning  UpdateFailed           69s                seldon-controller-manager  Failed to update status for SeldonDeployment "router": Operation cannot be fulfilled on seldondeployments.machinelearning.seldon.io "router": the object has been modified; please apply your changes to the latest version and try again
  Normal   CreateService          29s                seldon-controller-manager  Created Service "router-test"
  Normal   CreateVirtualService   29s                seldon-controller-manager  Created VirtualService "router-http"
  Normal   CreateVirtualService   29s                seldon-controller-manager  Created VirtualService "router-grpc"
  Normal   CreateDestinationRule  29s                seldon-controller-manager  Created DestinationRule "router-test"
  Normal   Updated                28s (x6 over 69s)  seldon-controller-manager  Updated SeldonDeployment "router"

**Comment by cliveseldon** [comment]
Usually these events can be ignored as they will be retried. Apart from this is everything working ok?

**Comment by pjpan** [comment]
> Usually these events can be ignored as they will be retried. Apart from this is everything working ok?

No, when I use python wrapper package router, router didn't work. the request will go direct  to the model node.

// seldon-deployment: 
> apiVersion: machinelearning.seldon.io/v1
kind: SeldonDeployment
metadata:
  name: router
spec:
  name: routertestr
  protocol: seldon
  annotations:
    seldon.io/rest-timeout: "25000"
  predictors:
    - componentSpecs:
        - spec:
            containers:
              - image: hrb.nioint.com/power/base/seldon/seldonio/custom-router-server:v1.0.0
                name: router
                env:
                - name: NUMBER_OF_ROUTERS
                  value: "3"
      graph:
        name: router
        implementation: CUSTOM_NODE_SERVER
        type: ROUTER
        modelUri: s3://com-nio-pow-data-kuafu-test/seldon_models/inference_graph/router
        envSecretRefName: seldon-rclone-secret
        children:
          - name: crack-detect
            implementation: TENSORFLOW_SERVER_CUSTOM
            modelUri: s3://com-nio-pow-data-kuafu-test/seldon_models/battery-surface-crack-cv-cdn
            envSecretRefName: seldon-rclone-secret
            parameters:
              - name: model_type
                type: STRING
                value: pytorch
            children: [ ]
          - name: gps-detect
            implementation: TENSORFLOW_SERVER_CUSTOM
            modelUri: s3://com-nio-pow-data-kuafu-test/seldon_models/battery-surface-gps-cv-cdn
            envSecretRefName: seldon-rclone-secret
            parameters:
              - name: model_type
                type: STRING
                value: pytorch
            children: [ ]
          - name: image-detect
            implementation: CUSTOM_ALGO_SERVER
            modelUri: s3://com-nio-pow-data-kuafu-test/ai-models/battery-surface-image-cv-cdn/v0.0.1
            envSecretRefName: seldon-rclone-secret
            children: [ ]
      name: test
      replicas: 1


// CustomRouterServer.py 

```
import logging
import random
import os
import sys
import numpy as np
log = logging.getLogger(__name__)

NUMBER_OF_ROUTERS = int(os.environ.get("NUMBER_OF_ROUTERS", "2"))
log.info(f"number of routes: {NUMBER_OF_ROUTERS}")

CUSTOM_ROUTER = "custom_router.py" 
MODEL_PATH = "/mnt/models"
ROUTE_PATH = os.path.join(MODEL_PATH, CUSTOM_ROUTER)

class CustomRouterServer(object):
    def __init__(self, model_uri=None):
        super().__init__()
        self.model_uri = model_uri
        self.ready = False
        self._custom_router = None
        self.load()

    def load(self): 
        """
        :return:
        """
        if os.path.exists(MODEL_PATH):
            sys.path.append(MODEL_PATH)
            log.info(f"model path:{MODEL_PATH} append success, current version is {__version__}")
        else:
            # os.makedirs(MODEL_PATH)
            log.info(f"{MODEL_PATH} does not exists")

        # 判断router是否存在，存在则进行模块加载
        if os.path.exists(ROUTE_PATH):
            from custom_router import custom_router
            self._custom_router = custom_router
            log.info(f"custom router loading success")

        self.ready = True

    def route(self, features):
        """
        用来做路由，根据模型的不同输入判断条件来进行路由
        :param features: 输入特征
        :param names:
        :param meta:
        :return: int，返回定向到哪一个节点的结果
        """
        if not self.ready:
            self.load()
        log.info(f"[SS] received features is : {features}")
        if os.path.exists(ROUTE_PATH):
            route = self._custom_router(features, NUMBER_OF_ROUTERS)
        else:
            route = random.randint(0, NUMBER_OF_ROUTERS)  # 随机分流
        log.info(f"[SR] routing to: {route}")

        return route
```

// Dockerfile.  SERVICE_TYPE=MODEL/ROUTER both didn't work.
```
FROM seldonio/seldon-core-s2i-python3:1.13.1
WORKDIR /app
# Install opencv dependency packages
RUN yum update -y &&yum -y  install mesa-libGL
# Install python packages
COPY requirements.txt requirements.txt
RUN pip install -r requirements.txt -i https://mirrors.aliyun.com/pypi/simple/
# Copy source code
COPY . .
# Port for GRPC
EXPOSE 5000
# Port for REST
EXPOSE 9000
# Define environment variables
ENV MODEL_NAME=CustomRouterServer SERVICE_TYPE=MODEL
CMD exec seldon-core-microservice $MODEL_NAME --service-type $SERVICE_TYPE 
```
